store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/util/MergeSchema.java

```diff
@@ -48,20 +60,52 @@ public Schema apply(final Schema update, final Schema state) {
         if (state == null) {
             return nonNull(update) ? update : new Schema();
         } else {
-            Schema.Builder mergeSchema = new Schema.Builder(state);
             //Check if Vertex Serialiser needs to be wiped due to previous clash with merging.
             final Serialiser vertexSerialiser = (boolean) context.getOrDefault(WIPE_VERTEX_SERIALISERS, false)
                     ? null
                     : update.getVertexSerialiser();
+
+            //Check if visibility property needs to be wiped due to previous clash with merging.
+            final String visibilityProperty = (boolean) context.getOrDefault(WIPE_VISIBILITY_PROPERTY, false)
+                    ? null
+                    : update.getVisibilityProperty();
+
+            Schema.Builder mergeSchema = new Schema.Builder(state);
+
             try {
-                mergeSchema.merge(new Schema.Builder(update).vertexSerialiser(vertexSerialiser).build());
-            } catch (final SchemaException e) {
-                if (e.getMessage().contains(Schema.UNABLE_TO_MERGE_SCHEMAS_CONFLICT_WITH_VERTEX_SERIALISER_OPTIONS_ARE)) {
-                    LOGGER.error(String.format("Caught SchemaException, attempting to re-merge but without vertex serialisers. Error message:%s", e.getMessage()));
+                mergeSchema.merge(new Schema.Builder(update)
+                        .vertexSerialiser(vertexSerialiser)
+                        .visibilityProperty(visibilityProperty)
+                        .build());
+
+            } catch (final Exception e) {
+                if (e.getMessage().contains(SCHEMAS_CONFLICT_WITH_VERTEX_SERIALISER)) {
```

Comment:
Maybe rather than all these `if (e.getMessage().contains(CONSTANT)`, there could be different types of schema exception that are caught?

Reply:
I'm not sure I agree with this. I don't want to create 4-6 new Exception classes just for Info regarding a particular type of SchemaException.

Reply:
Having multiple exception classes is pretty widely accepted as the 'correct' way to do exception handling in java over things like return codes as you can then catch and handle each one as needed with the `catch` block to give a clear program flow e.g.
```java
} catch (final CustomExceptionOne e1) {
...
} catch (final CustomExceptionTwo e2) {
...
} catch (final Exception e3) {
...
}
```
You could also enforce custom formatting or params for your exceptions to reduce some of the extra string formatting you are doing when throwing them.

Reply:
Adding the 4-6 new Exception Classes I don't believe will help because of re-throwing. 
 I don't receive the same type of class that was thrown. I would have to create new classes and then navigate down the `e.getCause()` path.

See: 
[Schema ](https://github.com/gchq/Gaffer/blob/f63eca2053be7ecf592a7899c16db1fa1c21d239/core/store/src/main/java/uk/gov/gchq/gaffer/store/schema/Schema.java#L430)

https://github.com/gchq/Gaffer/blob/f63eca2053be7ecf592a7899c16db1fa1c21d239/core/store/src/main/java/uk/gov/gchq/gaffer/store/schema/Schema.java#L355

Reply:
at [TypeDef L170](https://github.com/gchq/Gaffer/blob/f63eca2053be7ecf592a7899c16db1fa1c21d239/core/store/src/main/java/uk/gov/gchq/gaffer/store/schema/TypeDefinition.java#L170) I am currently throwing a SchemaException even though I'm only merging TypeDefinitions.

maybe I could remove Schema L355 and Schema L430 and provide more information at TypeDef 170.

Reply:
please examine https://github.com/gchq/Gaffer/pull/3051 as to why I don't think it will work.

I loose adding information about types, and I am unable to catch it in the MergeSchema.

Reply:
https://github.com/gchq/Gaffer/pull/3051

has now been improved and satisfies the concerns I had about untidy code and re-throwing many similar exceptions.